queues/rabbitmq/rabbitmq.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	return &Transport{
+		sendChans:    make(map[string]chan []byte),
+		receiveChans: make(map[string]chan []byte),
+		errChan:      make(chan error, 10),
+		stopchan:     make(chan struct{}),
+		stopPubChan:  make(chan struct{}),
+		stopSubChan:  make(chan struct{}),
+		connection:   options.Connection,
+	}
+}
+
+func (t *Transport) newConnection() (*amqp.Connection, error) {
+	var err error
+	if t.connection != nil {
+		return t.connection, nil
+	}
+
+	conn, err := amqp.Dial(DefaultAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	t.connection = conn
+	return t.connection, err
+}
+
+// Receive gets a channel on which to receive messages
+// with the specified name.
+func (t *Transport) Receive(name string) <-chan []byte {
+	t.Lock()
+	defer t.Unlock()
+
+	ch, ok := t.receiveChans[name]
+	if ok {
+		return ch
+	}
+
+	ch, err := t.makeSubscriber(name)
+	if err != nil {
+		t.errChan <- vice.Err{Name: name, Err: err}
+		return make(chan []byte)
+	}
+
+	t.receiveChans[name] = ch
+	return ch
+}
+
+func (t *Transport) makeSubscriber(name string) (chan []byte, error) {
+	c, err := t.newConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	rmqch, err := c.Channel()
+	if err != nil {
+		return nil, err
+	}
+
+	q, err := rmqch.QueueDeclare(
+		name,  // name
+		true,  // durable
+		false, // delete when used
+		false, // exclusive
+		false, // no-wait
+		nil,   // arguments
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// fair load balancing messages on queue when there are multiple receivers
+	err = rmqch.Qos(
+		1,     // prefetch count
+		0,     // prefetch size
+		false, // global
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	msgs, err := rmqch.Consume(
+		q.Name, // queue
+		"",     // consumer
+		false,  // auto-ack
+		false,  // exclusive
+		false,  // no-local
+		false,  // no-wait
+		nil,    // args
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan []byte, 1024)
+	go func() {
+		defer rmqch.Close()
+		for {
+			select {
+			case d := <-msgs:
+				ch <- d.Body
+				d.Ack(false)
+			case <-t.stopSubChan:
+				return
+			}
+		}
+	}()
+	return ch, nil
+}
+
+// Send gets a channel on which messages with the
+// specified name may be sent.
+func (t *Transport) Send(name string) chan<- []byte {
+	t.Lock()
+	defer t.Unlock()
+
+	ch, ok := t.sendChans[name]
+	if ok {
+		return ch
+	}
+
+	ch, err := t.makePublisher(name)
+	if err != nil {
+		t.errChan <- vice.Err{Name: name, Err: err}
+		return make(chan []byte)
+	}
+	t.sendChans[name] = ch
+	return ch
+}
+
+func (t *Transport) makePublisher(name string) (chan []byte, error) {
+	c, err := t.newConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	rmqch, err := c.Channel()
+	if err != nil {
+		return nil, err
+	}
+
+	q, err := rmqch.QueueDeclare(
+		name,  // name
+		true,  // durable
+		false, // delete when unused
+		false, // exclusive
+		false, // no-wait
+		nil,   // arguments
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan []byte, 1024)
```

Comment:
Have you benchmarked the transport with and without this buffered channel?
We could remove the cleanup logic if the performance is the same.

Reply:
Thanks for the review. Typos are fixed. I've changed the buffered channel to unbuffered as you suggested because performance is about the same.